Add float type caster and revert type hint changes to int_ and float_

gentlegiantJGC opened the pull request:
## Description

The casting rules and type hints for `py::float_` and `py::int_` should match the python type hint conventions.
An input typed with `int` can only accept `int` types and `float` can accept `float` or `int` types as defined in as per PEP 484.
`when an argument is annotated as having type float, an argument of type int is acceptable`

Expected behaviour
| Class | can cast from | type hint |
| ----- | --------------- | ---------- |
| py::int_ | int | int |
| py::float_ | float or int | float |

Current behavour
| Class | can cast from | type hint |
| ----- | --------------- | ---------- |
| py::int_ | int | typing.SupportsInt |
| py::float_ | float | typing.SupportsFloat |

The type hints were changed in #5540 by @InvincibleRMC but the classes do not support these casting rules.

```cpp
m.def("test_int", [](py::int_ i) { return i; });
m.def("test_float", [](py::float_ f) { return f; });
```

```py
test_int(5.5)
```
```
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: test_int(): incompatible function arguments. The following argument types are supported:
    1. (arg0: typing.SupportsInt) -> int
Invoked with: 5.5
```
```py
test_float(5)
```
```
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: test_float(): incompatible function arguments. The following argument types are supported:
    1. (arg0: typing.SupportsFloat) -> float
Invoked with: 5
```

Fixes #5840

## Suggested changelog entry:

* Fix py::float_ casting and py::int_ and py::float_ type hints

gentlegiantJGC commented:
I feel like `py::float_` should handle automatically casting from a python int as per PEP 484
`when an argument is annotated as having type float, an argument of type int is acceptable`
But that should be another pull request.

rwgk commented on tests/test_pytypes.cpp:
```diff
@@ -209,6 +209,7 @@ TEST_SUBMODULE(pytypes, m) {
     m.def("get_tuple_from_iterable", [](const py::iterable &iter) { return py::tuple(iter); });
     // test_float
     m.def("get_float", [] { return py::float_(0.0f); });
+    m.def("cast_float", [](py::float_ f) { return f; });
```
Could you please rename this to `roundtrip_float` or `float_roundtrip` or similar?

(`cast` is super ambiguous in the context pybind11; C/C++ cast, to-Python, from-Python, all in the same name bucket unfortunately)

gentlegiantJGC replied:
done

rwgk reviewed (commented):
Looks good to me. @InvincibleRMC could you please also review & formally approve? I'll merge this PR after I see your approval.

gentlegiantJGC commented:
@rwgk do you have any idea what is causing the tests to fail?
I can't see what I have changed that would cause the segfault

rwgk commented:
I'm using PR #3939 to see if current `master` has the same issue with the two graalpy tests.

rwgk commented:
> > I'm using PR #3939 to see if current `master` has the same issue with the two graalpy tests.
> 
> Looks like it only ran half the tests and not the ones that failed here.

Yeah :-(
I forgot about that. I converted to Ready for review to trick it into running everything.

rwgk commented:
I just manually cancelled the workflow, to not block those two runner slots for several more hours until the tests hit the timeout.

gentlegiantJGC commented:
The tests that failed here have passed in your PR. It must be something I changed but I have no idea what it is.

The output is the same up to this line
```
 ../../tests/test_pytypes.py::test_surrogate_pairs_unicode_error[str_from_std_string_input] XFAIL [ 76%]
```

This is the first test that did not run
```
../../tests/test_pytypes.py::test_bytes PASSED                           [ 76%]
```

Are your tests run sequentially? Is this where it is failing?

The only potential break I can see is the removal of `handle_type_name` for `float_`.
I thought it was not needed with it added to the custom `type_caster` but perhaps something needs it?

rwgk commented:
Trying another super quick and easy thing: PR #5850

Just started. But now I see already that the total number of tests is 85, like here.

rwgk commented:
All 83 tests pass under #5850: it seems very certain now that there is a problem with this PR, unfortunately, or maybe this PR is triggering a latent bug in graalpy.

@gentlegiantJGC I think it could save you time (potentially a lot) if you wait until we hear back from @msimacek.

gentlegiantJGC commented:
It was a reference counting mistake I had made.
I have fixed that and simplified the implementation a bit.

Another test is now stuck indefinetly. No error for this one though.

You may want to add a more sensible maximum time limit for workflows.
The default of 6 hours is a bit high.
https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#jobsjob_idtimeout-minutes

rwgk commented:
> Another test is now stuck indefinetly. No error for this one though.

I think I have seen that one a few times before, most likely it's unrelated to this PR. I just cancelled the job and then triggered a rerun.

> You may want to add a more sensible maximum time limit for workflows. The default of 6 hours is a bit high. https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#jobsjob_idtimeout-minutes

It'd be awesome if you could send a PR for that. 90 minutes will probably never time out unless there is an actual bug.
